guides/keras_cv/object_detection_keras_cv.py

```diff
@@ -0,0 +1,868 @@
+    classification_loss="focal",
+    box_loss="smoothl1",
+)
+
+"""
+### Metric Evaluation
+
+Just like any other metric, you can pass the `KerasCV` object detection metrics
+to `compile()`.  The most popular Object Detection metrics are COCO metrics,
+which were published alongside the MSCOCO dataset.  KerasCV provides an easy
+to use suite of COCO metrics. under the `keras_cv.metrics.BoxCOCOMetrics`
+symbol:
+"""
+
+coco_metrics = keras_cv.metrics.BoxCOCOMetrics(
+    bounding_box_format="xywh", evaluate_freq=128
+)
+
+"""
+Lets define a quick helper to print our metrics in a nice table:
+"""
+
+def print_metrics(metrics):
+  maxlen = max([len(key) for key in result.keys()])
+  print("Metrics:")
+  print("-" * (maxlen+1))
+  for k, v in metrics.items():
+      print(f"{k.ljust(maxlen+1)}: {v.numpy():0.2f}")
+
+"""
+Due to the high computational cost of computing COCO metrics, the KerasCV
+`BoxCOCOMetrics` component requires an `evaluate_freq` parameter to be passed to
+its constructor.  Every `evaluate_freq`-th call to `update_state()`, the metric
+will recompute the result.  In between invocations, a cached version of the
+result will be returned.
+
+To force an evaluation, you may call `coco_metrics.result(force=True)`:
+"""
+
+pretrained_model.compile(
+    classification_loss="focal",
+    box_loss="smoothl1",
+    optimizer=optimizer,
+    metrics=[coco_metrics],
+)
+coco_metrics.reset_state()
+result = pretrained_model.evaluate(eval_ds.take(40), verbose=0)
+result = coco_metrics.result(force=True)
+
+print_metrics(result)
+
+"""
+
+**A note on TPU compatibility:**
+
+Evaluation of `BoxCOCOMetrics` require running `tf.image.non_max_suppression()`
+inside of the `model.train_step()` and `model.evaluation_step()` functions.
+Due to this, the metric suite is not compatible with TPU when used with the
+`compile()` API.
+
+Luckily, there are two workarounds that allow you to still train a RetinaNet on TPU:
+
+- The use of a custom callback
+- Using a [SideCarEvaluator](https://www.tensorflow.org/api_docs/python/tf/keras/utils/SidecarEvaluator)
+
+Let's use a custom callback to achieve TPU compatibility in this guide:
+"""
+
+class EvaluateCOCOMetricsCallback(keras.callbacks.Callback):
+    def __init__(self, data):
+        super().__init__()
+        self.data = data
+        self.metrics = keras_cv.metrics.BoxCOCOMetrics(
+            bounding_box_format="xywh",
+            # passing 1e9 ensures we never evaluate until
+            # `metrics.result(force=True)` is
+            # called.
+            evaluate_freq=1e9,
+        )
+
+    def on_epoch_end(self, epoch, logs):
+        self.metrics.reset_state()
+        for batch in tqdm.tqdm(self.data):
+            images, y_true = batch[0], batch[1]
+            y_pred = self.model.predict(images, verbose=0)
+            self.metrics.update_state(y_true, y_pred)
+
+        metrics = self.metrics.result()
+        logs.update(metrics)
+        return logs
+
+"""
+## Training our model
```

Comment:
I think this is the penultimate section of the guide. It's weird to go back and forth between the pretrained and random models.

Reply:
I see what you are saying here, but I was struggling to come up with a better solution.